File: app/src/main/java/oblitusnumen/calendar/implementation/data/Entry.java
package oblitusnumen.calendar.implementation.data;


import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.database.Cursor;
import android.provider.BaseColumns;
import oblitusnumen.calendar.implementation.Utils;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Entry implements BaseColumns {
    public static final String TABLE_NAME = "entries";
    public static final String COLUMN_NAME_ID = "id";
    public static final String COLUMN_NAME_NAME = "name";
    public static final String CONTENTS_FILENAME = "contents.md";
    private final DbManager dbManager;
    int id = -1;
    String name;

    @SuppressLint("NewApi")
    public String getContents() {
        try (FileInputStream fis = new FileInputStream(getContentsFile())) {
            return new String(fis.readAllBytes());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public List<Date> getDates() {
        List<Date> dates = new ArrayList<>();
        try (Cursor query = dbManager.getReadableDatabase().rawQuery("SELECT * FROM " + Date.TABLE_NAME + " WHERE " + Date.COLUMN_NAME_ENTRY_ID + " = ?",
                new String[]{String.valueOf(id)})) {
            if (query != null) {
                while (query.moveToNext()) {
                    dates.add(new Date(dbManager, query));
                }
            }
        }
        return dates;
    }

    Entry(DbManager dbManager, ContentValues contentValues) {
        this.dbManager = dbManager;
        this.id = (int) contentValues.get(COLUMN_NAME_ID);
        this.name = (String) contentValues.get(COLUMN_NAME_NAME);
    }

    @SuppressLint("Range")
    Entry(DbManager dbManager, Cursor cursor) {
        this(dbManager, cursor.getInt(cursor.getColumnIndex(COLUMN_NAME_ID)), cursor.getString(cursor.getColumnIndex(COLUMN_NAME_NAME)));
    }

    /**
     * we initialize <code>Entry</code> here and only here
     * @param dbManager
     */
    Entry(DbManager dbManager) {
        this.dbManager = dbManager;
        ContentValues contentValues = toContentValues();
        contentValues.put(COLUMN_NAME_ID, (Integer) null);
        id = (int) dbManager.getWritableDatabase().insert(TABLE_NAME, null, contentValues);
        if (!getDir().mkdirs()) throw new RuntimeException("could not create directory for entry " + id + ", filename: " + getDir());
        try {
            if (getContentsFile().createNewFile()) throw new IOException();
        } catch (IOException e) {
            throw new RuntimeException("could not create directory for entry " + id + ", filename: " + getDir(), e);
        }
    }

    public File getDir() {
        return new File(dbManager.activity.getFilesDir(), String.valueOf(id));
    }

    public File getContentsFile() {
        return new File(getDir(), CONTENTS_FILENAME);
    }

    public void delete() {
        dbManager.getWritableDatabase().execSQL("DELETE FROM " + TABLE_NAME + " WHERE " + COLUMN_NAME_ID + " = ?", new String[]{String.valueOf(id)});
        Utils.rmRecursively(getDir());
        dbManager.getWritableDatabase().execSQL("DELETE FROM " + EntryTagLinks.TABLE_NAME + " WHERE " + EntryTagLinks.COLUMN_NAME_ENTRY_ID + " = ?", new String[]{String.valueOf(id)});
//        dbManager.getWritableDatabase().execSQL("DELETE FROM " + Tag.TABLE_NAME + " WHERE " + Tag.COLUMN_NAME_ENTRY_ID + " = ?", new String[]{String.valueOf(id)});
        dbManager.getWritableDatabase().execSQL("DELETE FROM " + Date.TABLE_NAME + " WHERE " + Date.COLUMN_NAME_ENTRY_ID + " = ?", new String[]{String.valueOf(id)});
        dbManager.getWritableDatabase().execSQL("DELETE FROM " + Notification.TABLE_NAME + " WHERE " + Notification.COLUMN_NAME_ENTRY_ID + " = ?", new String[]{String.valueOf(id)});

        // FIXME: 10/24/24 remove all asociated entries i.e.
    }

    private Entry(DbManager dbManager, int id, String name) {
        this.dbManager = dbManager;
        this.id = id;
        this.name = name;
    }

    ContentValues toContentValues() {
        ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_NAME_ID, id);
        contentValues.put(COLUMN_NAME_NAME, name);
        return contentValues;
    }

    public String getName() {
        return name;
    }

    public List<Tag> getTags() {
        List<Tag> tags = new ArrayList<>();
        try (Cursor cursor = dbManager.getReadableDatabase().rawQuery("SELECT " + Tag.TABLE_NAME + ".* " +
                        "FROM " + Tag.TABLE_NAME + " " +
                        "JOIN " + EntryTagLinks.TABLE_NAME + " ON " + Tag.TABLE_NAME + "." + Tag.COLUMN_NAME_ID + " = " + EntryTagLinks.TABLE_NAME + "." + EntryTagLinks.COLUMN_NAME_TAG_ID + " " +
                        "WHERE " + EntryTagLinks.TABLE_NAME + "." + EntryTagLinks.COLUMN_NAME_ENTRY_ID + " = ?",
                new String[]{String.valueOf(id)})) {
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    tags.add(new Tag(dbManager, cursor));
                }
            }
        }
        return tags;
    }

    public void set(@NotNull String name, @NotNull List<Tag> tags, @NotNull List<Date> dates, @NotNull String contents) {
        try (FileOutputStream fos = new FileOutputStream(getContentsFile())) {
            fos.write(contents.getBytes());
        } catch (IOException e) {
            throw new RuntimeException("could not save contents file for entry " + id, e);
        }

        this.name = name;
        update();

        LinkedList<Tag> tags1 = new LinkedList<>(tags);
        tags1.removeIf(t -> {
            if (t.id == -1) {
                t.create();
                addTag(t.id);
                return true;
            } return false;
        });
//        Map<Integer, Tag> tagsNew = tags1.stream().collect(Collectors.toMap(t -> t.id, t -> t));
        Map<Integer, Tag> tagsOld = getTags().stream().collect(Collectors.toMap(t -> t.id, t -> t));
        for (Tag tag : tags1) {
            if (tagsOld.containsKey(tag.id)) {
                tagsOld.remove(tag.id);
            } else {
                addTag(tag.id);
            }
        }
        for (Integer t : tagsOld.keySet()) {
            rmTag(t);
        }

        LinkedList<Date> dates1 = new LinkedList<>(dates);
        dates1.removeIf(t -> {
            if (t.id == -1) {
                t.createOrUpdate();
                return true;
            } return false;
        });
//        Map<Integer, Date> datesNew = dates1.stream().collect(Collectors.toMap(t -> t.id, t -> t));
        Map<Integer, Date> datesOld = getDates().stream().collect(Collectors.toMap(t -> t.id, t -> t));
        for (Date date : dates1) {
            if (datesOld.containsKey(date.id)) {
                datesOld.remove(date.id);
            } else {
                date.create();
            }
        }
        for (Date d : datesOld.values()) {
            d.delete();
        }
    }

    void update() {
        dbManager.getWritableDatabase().execSQL("UPDATE " + TABLE_NAME + " " +
                "SET " + COLUMN_NAME_NAME + " = ?", new String[]{name});
    }

    void addTag(int tagId) {
        new EntryTagLinks(dbManager, id, tagId).create();
    }

    void rmTag(int tagId) {
        new EntryTagLinks(dbManager, id, tagId).delete();
    }
}

File: app/src/main/java/oblitusnumen/calendar/implementation/data/Entry.kt
package oblitusnumen.calendar.implementation.data

import android.content.ContentValues
import android.database.Cursor
import android.provider.BaseColumns
import androidx.compose.ui.graphics.Color
import androidx.core.database.sqlite.transaction
import oblitusnumen.calendar.implementation.rmRecursively
import oblitusnumen.calendar.implementation.toColor
import oblitusnumen.calendar.implementation.toInt
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

class Entry private constructor(
    private val dbManager: DbManager,
    id: Int? = null,
    private var state: Int = STATE_NEW,
    var name: String = "",
    var excludeFromCalendarView: Boolean = false,
    var color: Color? = null
) : BaseColumns {
    var id: Int? = id
        private set

    fun getColorOrDefault(): Color = color ?: dbManager.defaultEntryColor

    fun isNotCreated() = id == null

    private fun create() {
        val contentValues = getContentValues()
        contentValues.put(COLUMN_NAME_ID, null as Int?)
        id = dbManager.writableDatabase.insert(TABLE_NAME, null, contentValues).toInt()
        if (!dir.mkdirs() || !contentsFile.createNewFile()) {
            this.id = null
            try {
                if (dir.exists()) rmRecursively(dir)
                dbManager.writableDatabase.execSQL(
                    "DELETE FROM $TABLE_NAME WHERE $COLUMN_NAME_ID = ?",
                    arrayOf(id.toString())
                )
            } catch (_: Exception) {
            }
            throw IOException("could not setup directory for entry $id, filename: $dir")
        }
    }

    private fun tryUpdateContents(contents: String) {// FIXME: atomic saves
        try {
            FileOutputStream(contentsFile).use { fos ->
                fos.write(contents.toByteArray())
                fos.flush()
                fos.fd.sync()
            }
        } catch (e: IOException) {
            throw IOException("could not save contents for entry $id", e)
        }
    }

    private fun update() {
        dbManager.writableDatabase.update(TABLE_NAME, getContentValues(), "$COLUMN_NAME_ID = ?", arrayOf(id.toString()))
    }

    @Throws(IOException::class)
    fun deleteCascade() {
        if (isNotCreated()) return
        dbManager.writableDatabase.transaction {
            state = STATE_DELETED
            update()
            dbManager.writableDatabase.execSQL(
                "DELETE FROM ${EntryTagLinks.TABLE_NAME} WHERE ${EntryTagLinks.COLUMN_NAME_ENTRY_ID} = ?",
                arrayOf(id.toString())
            )
            dbManager.writableDatabase.execSQL(
                "DELETE FROM ${Notification.TABLE_NAME} WHERE ${Notification.COLUMN_NAME_ENTRY_ID} = ?",
                arrayOf(id.toString())
            )
            dbManager.writableDatabase.execSQL(
                "DELETE FROM ${Date.TABLE_NAME} WHERE ${Date.COLUMN_NAME_ENTRY_ID} = ?",
                arrayOf(id.toString())
            )
        }
        delete()
        id = null
        dbManager.tryScheduleNotification()
    }

    private fun delete() {
        if (dir.exists()) rmRecursively(dir)
        dbManager.writableDatabase.execSQL(
            "DELETE FROM $TABLE_NAME WHERE $COLUMN_NAME_ID = ?",
            arrayOf(id.toString())
        )
    }

    fun fixup() {
        when (state) {// FIXME: maybe differ new and deleted
            STATE_NEW, STATE_DELETED -> {
                delete()
            }

            STATE_UPDATING -> {
                // FIXME: repair file system
                state = STATE_NORMAL
                update()
            }

            else -> throw IllegalStateException("unknown state $state")
        }
    }

    private fun getContentValues(): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(COLUMN_NAME_STATE, state)
        contentValues.put(COLUMN_NAME_NAME, name)
        contentValues.put(COLUMN_NAME_EXCLUDE_VIEW, if (excludeFromCalendarView) 1 else 0)
        contentValues.put(COLUMN_NAME_COLOR, color.toInt())
        return contentValues
    }

    private val dir: File
        get() = File(dbManager.filesDir, id.toString())

    private val contentsFile: File
        get() = File(dir, CONTENTS_FILENAME)

    fun getContents(): String { //todo contents should not be a just a string
        if (!contentsFile.exists()) return ""
        try {
            FileInputStream(contentsFile).use { fis ->
                val buffer = ByteArray(4096) // Buffer to hold file data
                val content = StringBuilder()
                var bytesRead: Int
                while ((fis.read(buffer).also { bytesRead = it }) != -1) {
                    content.append(String(buffer, 0, bytesRead)) // Convert bytes to string
                }
                return content.toString() // Return the complete content as a string
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun getNotifications(): List<Notification> {
        dbManager.readableDatabase.rawQuery(
            "SELECT * FROM ${Notification.TABLE_NAME} " +
                    "WHERE ${Notification.COLUMN_NAME_ENTRY_ID} = ?", arrayOf(id.toString())
        ).use { cursor ->
            return Notification.cursorToList(dbManager, cursor)
        }
    }

    fun getDates(): List<Date> {
        if (isNotCreated()) return emptyList()
        return Date.getAllByEntryId(dbManager, id!!)
    }

    fun getTags(): List<Tag> {
        dbManager.readableDatabase.rawQuery(
            "SELECT ${Tag.TABLE_NAME}.* FROM ${Tag.TABLE_NAME} JOIN ${EntryTagLinks.TABLE_NAME} " +
                    "ON ${Tag.TABLE_NAME}.${Tag.COLUMN_NAME_ID} = ${EntryTagLinks.TABLE_NAME}.${EntryTagLinks.COLUMN_NAME_TAG_ID} " +
                    "WHERE ${EntryTagLinks.TABLE_NAME}.${EntryTagLinks.COLUMN_NAME_ENTRY_ID} = ?",
            arrayOf(id.toString())
        ).use { cursor ->
            return Tag.cursorToList(dbManager, cursor)
        }
    }

    @Throws(IOException::class)
    fun set(
        name: String,
        excludeFromCalendarView: Boolean,
        color: Color?,
        tags: Iterable<Tag>,
        dates: Iterable<Date>,
        notifications: Iterable<Notification>,
        contents: String
    ) {
        if (isNotCreated()) {
            create()
        } else {
            state = STATE_UPDATING
            update()
        }
        tryUpdateContents(contents)
        dbManager.writableDatabase.transaction {
            //setting entry
            this@Entry.state = STATE_NORMAL
            this@Entry.name = name
            this@Entry.excludeFromCalendarView = excludeFromCalendarView
            this@Entry.color = color
            update()

            //setting tags
            val tagsNew = tags.map { it.id }.toSet()
            val tagsOld = getTags().groupingBy { it.id }.reduce { _, accumulator, _ -> accumulator }
            for (tId in tagsOld.keys) {
                if (!tagsNew.contains(tId)) rmTag(tId!!)
            }
            for (t in tags) {
                if (!tagsOld.containsKey(t.id)) {
                    t.createIfNotExists()
                    addTag(t.id!!)
                }
            }

            //setting dates
            val datesNew = dates.map { it.id }.toSet()
            val datesOld = getDates().groupingBy { it.id }.reduce { _, accumulator, _ -> accumulator }
            for (d in datesOld.values) {
                if (!datesNew.contains(d.id)) d.delete()
            }
            for (d in dates) {
                if (datesOld.containsKey(d.id)) d.update() else {
                    d.entryId = id
                    d.create()
                }
            }

            //setting notifications
            val notificationsNew = notifications.map { it.offset.toString() }.toSet()
            val notificationsOld =
                getNotifications().groupingBy { it.offset.toString() }.reduce { _, accumulator, _ -> accumulator }
            for (n in notificationsOld.values) {
                if (!notificationsNew.contains(n.offset.toString())) n.delete()
            }
            for (n in notifications) {
                if (notificationsOld.containsKey(n.offset.toString())) n.update() else {
                    n.entryId = id
                    n.create()
                }
            }
        }
        dbManager.tryScheduleNotification()
    }

    private fun addTag(tagId: Int) {
        EntryTagLinks.create(dbManager, id!!, tagId)
    }

    private fun rmTag(tagId: Int) {
        EntryTagLinks.delete(dbManager, id!!, tagId)
    }

    companion object {
        const val TABLE_NAME: String = "entries"
        const val COLUMN_NAME_ID: String = "id"
        const val COLUMN_NAME_STATE: String = "state"
        const val COLUMN_NAME_NAME: String = "name"
        const val COLUMN_NAME_EXCLUDE_VIEW: String = "excludeView"
        const val COLUMN_NAME_COLOR: String = "color"
        const val STATE_NORMAL: Int = 0
        const val STATE_NEW: Int = 1
        const val STATE_UPDATING: Int = 2
        const val STATE_DELETED: Int = 3
        const val CONTENTS_FILENAME: String = "contents.md"

        fun new(dbManager: DbManager): Entry {
            return Entry(dbManager)
        }

        fun cursorToList(
            dbManager: DbManager,
            cursor: Cursor
        ): MutableList<Entry> {
            val entries: MutableList<Entry> = ArrayList()
            val idxId = cursor.getColumnIndex(COLUMN_NAME_ID)
            val idxState = cursor.getColumnIndex(COLUMN_NAME_STATE)
            val idxName = cursor.getColumnIndex(COLUMN_NAME_NAME)
            val idxExcludeCalendarView = cursor.getColumnIndex(COLUMN_NAME_EXCLUDE_VIEW)
            val idxColor = cursor.getColumnIndex(COLUMN_NAME_COLOR)
            while (cursor.moveToNext())
                entries.add(
                    Entry(
                        dbManager,
                        cursor.getInt(idxId),
                        cursor.getInt(idxState),
                        cursor.getString(idxName),
                        cursor.getInt(idxExcludeCalendarView) != 0,
                        cursor.getInt(idxColor).toColor()
                    )
                )
            return entries
        }
    }
}